pkg/request/pool.go

```diff
@@ -216,39 +226,214 @@ func (rp *Pool) NextRequests(ctx context.Context) [][]byte {
 	return rawRequests
 }
 
-func (rp *Pool) RemoveRequests(requests ...string) error {
-	if atomic.LoadUint32(&rp.batchingEnabled) == 0 {
-		rp.pending.RemoveRequests(requests...)
+func (rp *Pool) RemoveRequests(requestsIDs ...string) error {
+	if !rp.isBatchingEnabled() {
+		rp.pending.RemoveRequests(requestsIDs...)
 		return nil
 	}
 
-	for _, requestID := range requests {
+	for _, requestID := range requestsIDs {
 		rp.batchStore.Remove(requestID)
 	}
 	return nil
 }
 
-// Close removes all the requests, stops all the timeout timers.
+func (rp *Pool) Prune(predicate func([]byte) error) { // TODO pruning after they are batched already might be an issue
+	requestsToRemove := make([]string, 0, rp.options.MaxSize)
+	if rp.isBatchingEnabled() {
+		rp.batchStore.ForEach(func(_, v interface{}) {
+			req := v.(*requestItem).request
+			if predicate(req) != nil {
+				requestsToRemove = append(requestsToRemove, rp.inspector.RequestID(req).ID)
+			}
+		})
+	} else {
+		requests := rp.pending.GetAllRequests(rp.options.MaxSize)
+		for _, req := range requests {
+			if predicate(req) != nil {
+				requestsToRemove = append(requestsToRemove, rp.inspector.RequestID(req).ID)
+			}
+		}
+	}
+	err := rp.RemoveRequests(requestsToRemove...)
+	if err != nil {
+		rp.logger.Errorf("Couldn't remove pruned requests; error: %s", err)
+	}
+}
+
+// ResetPool resets the pool options
+func (rp *Pool) ResetPool(options PoolOptions, batching bool) {
+	defer atomic.StoreUint32(&rp.stopped, 0)
+
+	rp.Halt()
+
+	if options.MaxSize < rp.options.MaxSize {
+		rp.logger.Panicf("The new max size is smaller than the original max size")
+	}
+
+	requests := make([][]byte, 0, rp.options.MaxSize)
+	batchingWasEnabled := rp.isBatchingEnabled()
+	if batchingWasEnabled {
+		rp.batchStore.ForEach(func(_, v interface{}) {
+			requests = append(requests, v.(*requestItem).request)
+		})
+	} else {
+		rp.pending.Close()
+		requests = rp.pending.GetAllRequests(rp.options.MaxSize)
+	}
+
+	rp.batchStore = nil
+	rp.pending = nil
```

Comment:
don't we need to close the pending first? Otherwise we have a reference from the goroutine that runs `changeEpochs()` to the object, no?

Reply:
That is why I call `rp.Halt()` at the beginning of the function